### PR TITLE
Support generational ZGC

### PR DESCRIPTION
**Pull request by loyispa**
Hi, here is a enhance for Generational ZGC , towards https://github.com/microsoft/gctoolkit/issues/364

major changes:
1. Some events(PauseMark,ConcurrentMark, etc.) occur twice in a single Cycle,  using their first timestamp.
2. Sum the event durations and memory occupancies (live/garbage/allocated) for the young generation and old generation.

### CODE REVIEW

File: parser/src/main/java/com/microsoft/gctoolkit/parser/ZGCParser.java [hunk 1]
```diff
@@ -323,6 +340,7 @@ private class ZGCForwardReference {
         private ReclaimSummary reclaimed;
         private ReclaimSummary memorySummary;
         private ZGCMetaspaceSummary metaspace;
+        private int memoryScope; // 0: default all heap; 1: young generation; 2: old generation
```
Comment (karianna) [inline parser/src/main/java/com/microsoft/gctoolkit/parser/ZGCParser.java:343]: I'm thinking enum here?

File: parser/src/main/java/com/microsoft/gctoolkit/parser/ZGCParser.java [hunk 2]
```diff
@@ -363,107 +381,150 @@ ZGCCycle toZGCCycle(DateTimeStamp endTime) {
...
         public void setPauseMarkEndStart(DateTimeStamp pauseMarkEndStart) {
+            if (this.pauseMarkEndStart != null) {
+                return;
+            }
             this.pauseMarkEndStart = pauseMarkEndStart;
         }
 
         public void setPauseMarkEndDuration(double pauseMarkEndDuration) {
-            this.pauseMarkEndDuration = pauseMarkEndDuration;
+            this.pauseMarkEndDuration += pauseMarkEndDuration;
```
Comment (karianna) [inline parser/src/main/java/com/microsoft/gctoolkit/parser/ZGCParser.java:402]: Is this GenZGC vs ZGC reasoning or fixing what you feel is a bug?

Comment (karianna) [reply parser/src/main/java/com/microsoft/gctoolkit/parser/ZGCParser.java:402]: Same with the other += changes.

**Review by karianna — COMMENTED** [review]
Couple of early comments, I'll leave the overall review to @kcpeppe et al.

File: parser/src/main/java/com/microsoft/gctoolkit/parser/ZGCParser.java (hunk 2 above)
Comment (loyispa) [reply parser/src/main/java/com/microsoft/gctoolkit/parser/ZGCParser.java:402]: Due to genZ dividing pauseMarkEndDuration into young and old duration, so sum them equals the original meaning.
```
[4.120s][info][gc,phases   ] GC(0) Y: Pause Mark Start (Major) 0.066ms
......
[4.155s][info][gc,phases   ] GC(0) Y: Pause Mark End 0.016ms
......
[4.182s][info][gc,phases   ] GC(0) O: Pause Mark End 0.014ms
......
```

File: parser/src/main/java/com/microsoft/gctoolkit/parser/ZGCParser.java (hunk 1 above)
Comment (loyispa) [reply parser/src/main/java/com/microsoft/gctoolkit/parser/ZGCParser.java:343]: Yes, it's much better than a magic number.

**Comment by kcpeppe** [comment]
Currently a single ZGC cycle is represented by a single event. For genZ I was looking at 2 events. I've thought of having an event per phase but I was concerned that this might lead to some downstream confusion with the analytics. Also, as you've mentioned, repeated phases per cycle is an issue. This may become even more confused depending on the log levels used by different projects. So, I'm leaning towards the single event per cycle but I'm not stuck on it. Are there reasons to create event per phase within a cycle that I'm missing?

**Comment by loyispa** [comment]
I'm inclined to think you're right to be concerned. Creating events per phase seems to break the "consensus" that each GCID represents a separate gc event. It may cause confusion in the counting of gc events.

**Comment by loyispa** [comment]
@kcpeppe   Do you have time to take a look at this PR?  Or if you have new plans for the genZ,  please let me know :)

**Comment by kcpeppe** [comment]
I think I'd like to see the events collapsed into one for a young and one for a tenured collection.

**Comment by kcpeppe** [comment]
sorry for dribbling thoughts here but I'm thinking that genZ should be it's own parser.

**Comment by kcpeppe** [comment]
This PR has caused me to reverse my thinking about how to integrate genZ into GCToolKit. I'd like to make a couple of mods and then merge.

**Pull request closed without merging**